File: ruby/prime-factors/prime_factors.rb
module PrimeRefinements

  refine Integer do
    def prime?
      return false if self < 2
      return self == 2 if (self % 2).zero?
      return self == 3 if (self % 3).zero?
      return true if self < 9
      5.step(Math.sqrt(self).floor, 2) do |num|
        return false if (self % num).zero?
      end
      true
    end
  end

end

class PrimeFactors

  using PrimeRefinements

  def self.for(num)
    return [] if num < 2
    return [num] if num.prime?
    prime_factors = []
    last_prime = 2
    while last_prime <= Math.sqrt(num)
      if num % last_prime == 0
        prime_factors << last_prime
        num = num / last_prime
      else
        last_prime += 1
        last_prime += 1 if last_prime.even?
        last_prime += 1 until last_prime.prime?
      end
    end
    prime_factors << num
  end

end

File: python/prime-factors/prime_factors.py
import math

def factors(value):
    if value < 2:
        return []
    if is_prime(value):
        return [value]
    
    prime_factors = []
    last_prime = 2
    while last_prime <= math.floor(math.sqrt(value)):
        if value % last_prime == 0:
            prime_factors.append(last_prime)
            value = value // last_prime
        else:
            last_prime += 1
            while not is_prime(last_prime):
                last_prime += 1
        
    prime_factors.append(value)

    return prime_factors


def is_prime(num):
    if num < 2:
        return False 
    if num % 2 == 0:
        return num == 2 
    if num % 3 == 0:
        return num == 3
    if num < 9:
        return True 
    for e in range(5, math.floor(math.sqrt(num)), 2):
        if num % e == 0:
            return False
    return True